src/metabase/upload.clj

```diff
@@ -170,18 +173,57 @@
     "unnamed_column"
     (u/slugify (str/trim raw-name))))
 
+(defn- is-pk?
+  [[col-name type]]
+  (and (#{"id" "pk"} (u/lower-case-en (name col-name)))
+       (isa? type ::int)))
+
+(defn- add-pk-column
+  "Adds driver-specific auto-increment and non-nil/PK constraints to the first ID/PK column it finds; otherwise prepends
+  a new ID column to the columns list."
+  [driver name-type-pairs]
+  (let [found-it   (atom false)
```

Comment:
Rather than use an atom, I would have expected this to use a `reduce` with `reduced` for short-circuiting

Reply:
I tried, I think this is quite tidier. I used a trick to treat the name-type-pairs as an ordered-map to remove the need for bookkeeping.

```clojure
(defn- add-pk-column
  "Adds driver-specific auto-increment and non-nil/PK constraints to the first ID/PK column it finds; otherwise prepends
  a new ID column to the columns list."
  [driver name-type-pairs]
  (if-let [pk-name (first (keep (fn [[n t]]
                                  (when (is-pk? n t) n))
                                name-type-pairs))]
    ;; we already have a PK, update it
    (assoc (ordered-map/ordered-map name-type-pairs) pk-name {:type ::pk :opts (driver/pk-options driver false)})
    ;; otherwise, prepend a new ID column
    (cons [:id {:type ::pk :opts (driver/pk-options driver true) :exclude-in-insert? true}]
          name-type-pairs)))
```